src/osd/ECCommon.h

```diff
@@ -489,7 +494,7 @@ struct ECCommon {
       const std::set<int> &want,      ///< [in] desired shards
       bool for_recovery,         ///< [in] true if we may use non-acting replicas
       bool do_redundant_reads,   ///< [in] true if we want to issue redundant reads to reduce latency
-      std::map<pg_shard_t, std::vector<std::pair<int, int>>> *to_read   ///< [out] shards, corresponding subchunks to read
+      read_request_t *read_request ///< [out] shard_reads, corresponding subchunks / other sub reads to read
```

Comment:
as you already changing the signature of this function, and as this is the only out parameter (except from the int status), maybe consider returning it directly (as a tl::expected)

Reply:
I have moved most of this into a different PR - but will address these comments there. https://github.com/ceph/ceph/pull/59729

Reply:
I had a look into using tl::expected for this and the result was messier and slower. 

I can change read_request from a pointer into a reference, which may help with some of your concerns, as it is never null.

Reply:
The reason for it being messier is that the read_request constructor requires a couple of arguments, which need to be passed around on the stack.  This is both messy and I expect results in slower compiled code.